lib/multiplexer/multiplexer.go

```diff
@@ -196,6 +203,26 @@ func (m *Mux) Serve() error {
 	}
 }
 
+// protocolListener returns a registered listener for Protocol proto
+// and is safe for concurrent access.
+func (m *Mux) protocolListener(proto Protocol) *Listener {
+	m.RLock()
+	defer m.RUnlock()
```

Comment:
@smallinsky There are a number of places that call the getters after Serve (`kube/proxy.TLSServer.Serve`, `service.TeleportProcess.initAuthService`, and multiple tests). The call order could certainly be changed in all those places.

I couldn't figure out a clean way to provide feedback when future iterations mistakenly call the getters after `Serve` (without changing the API too much for a bug fix). The fact that it takes two calls, one to create a `Mux` and another to register listeners is the core issue. As long as two calls are required, future iterations could call them out of order, creating a race condition without much feedback. As long as the getters are part of the API, they need to be safe for concurrent access. I thought about a few different designs for dealing with call order, but only the first three don't require locks:

1. Making callers create `mux.Listener` instances first and attach them to `Config`. This would prevent misuse and not require locks, but calling code would become more complicated. 
2. Briefly locking in `Serve` and copying registered listeners to local variables. This would still need to provide some feedback to developers if getters are called after Serve started.
3. Adding a `panic` to the getters with a good error message if called out of order, but it looked like `panic` was something we liked to avoid.
4. Logging a debug message and return `nil` from the getters so callers would panic, but didn't love relying on log messages to provide feedback.
5. Creating a new single method getter API that returned an error when called out of order: `mux.NewListener(Protocol) (net.Listener, error)`, but felt like it complicated things too much when the error will never occur if called in the correct order (and it still requires locks).

Number 1 makes it impossible to use the API incorrectly but UX for callers isn't great. Number 2 prevents the need for read locks in the `Serve/detectAndForward` loop, but requires 3, 4, or 5 in the getters. 3-5 don't really solve much, but do aim to provide decent feedback.

I decided to choose the path of least resistance and get some feedback first. Thoughts?

Reply:
All right. Thanks for explanation. It looks like that  `mux.Serve` call followed by `mux.DB/TLS/SSH` is abuse of multiplex library usage. 

Probably if we know that there will be only a few  `Write` calls during init handling and most of those calls will be  `Read` call during runtime the `atomic.Load` can be leverage here though let's lever the current version to keep things simple and avoid premature optimisation.

Reply:
Something about this still seems a bit off to me. I would expect the caller that creates a `Mux` to know which listeners they want to use. Could we leverage functional options here to allow a caller to do something like this: 

```go
mux, err := multiplexer.New(multiplexer.Config{
		Context:             t.Context,
		Listener:            listener,
		Clock:               t.Clock,
		EnableProxyProtocol: true,
		ID:                  t.Component,
	}, multiplexer.WithTLS())
```

By doing this we allow the callers to declare their intent and create the listeners they need up front. The rest of the API can remain the same and callers can still retrieve their individual listeners via the getters. If they forget to declare a particular listener in `New`, they will likely get a panic for using a nil listener during testing and fix the issue right away.

Reply:
@rosstimothy I get it. I went through multiple designs. My very first iteration changed the options from `Disable*` to `Enable*`, which is what the options in your example would essentially do. But we still have an issue with a listener being created by `Mux` and never serviced by the caller since there is no guarantee that a getter method is called. By explicitly creating listeners in the getters, we reduce the likelihood of a listener not being serviced.

As mentioned in the previous comment, we could force callers to create `mux.Listener` instances explicitly and pass them through the `New` call, but that seems like a lot of rework for questionable benefit. The logic to determine whether a protocol/listener is needed for all the different scenarios is not simple.

We need to ensure the API minimizes the scenario where the caller says they want to enable a listener but never actually services it. Reducing the steps to a single getter call seemed to solve the problem with little change to existing callers.

Reply:
> I would expect the caller that creates a Mux to know which listeners they want to use.

They do, but there is a mix of some callers repeating enable/disable logic in both `New` and when retrieving the listener. This could be updated, but the question is whether we want to do that for this security issue or in a refactor PR. 

> Could we leverage functional options

If we use flags, then I'd think we'd flip the old `Disable` flags to `Enable` in `multiplexer.Config`.

> If they forget to declare a particular listener in New, they will likely get a panic for using a nil listener during testing and fix the issue right away.

I was concerned with the scenario where a caller _enables_ a listener when they don't mean to and don't receive any feedback via test/runtime failure. This seemed more likely when they are repeating init logic in both `New` and when configuring the service that will use the listener. If they enable by mistake in `New` but correctly disable in their init logic, then a listener will be created and never serviced. Preventing this means trusting that when they request a listener they are attaching it in init logic and the listener will be serviced, which renders the enable/disable flags of questionable value.

@rosstimothy

Reply:
Yeah it makes a bit more sense now that I looked a bit more into it. Something about the multiplexer and listeners still seems like a clunky experience to me. But I think this PR solves the underlying issue and that could be addressed in the future.

Reply:
> Something about the multiplexer and listeners still seems like a clunky experience to me.

I agree.